File: 0189-rotate-array/0189-rotate-array.java
class Solution {
    public void rotatereversepart(int[] nums,int i,int j){
        int n=nums.length;
        while(i<=j){
            int temp=nums[i];
            nums[i]=nums[j];
            nums[j]=temp;
            i++;
            j--;
        }
    }
    public void rotate(int[] nums, int k) {
        int n=nums.length;

        k=k%n;
        if(k==0) return;

        rotatereversepart(nums,0,n-k-1);
        rotatereversepart(nums,n-k,n-1);
        rotatereversepart(nums,0,n-1);
    }
}

File: 0189-rotate-array/0189-rotate-array.cpp
class Solution {
public:
    void rotatepart(vector<int>& nums,int i,int j){
        int n=nums.size();
        while(i<=j){
            swap(nums[i],nums[j]);
            i++;
            j--;
        }
    }
    void rotate(vector<int>& nums, int k) {
        int n=nums.size();
        k=k%n;
        if(k==0) return ;
        rotatepart(nums,0,n-k-1);
        rotatepart(nums,n-k,n-1);
        rotatepart(nums,0,n-1);
    }
};